src/Plugin.php

```diff
@@ -52,28 +44,57 @@ public static function gm_get_shipping_page_link_return_string($text, $product,
       return $text;
     }
 
-    // If product level detail Alternative Shipping Information is not used then go through shop level rules.
-    $price = $product->get_price();
-    $previous_price_step = 0;
-    for ($i = 0; $i <= Plugin::LIMITS; $i++) {
-      $price_step = get_option('_' . Plugin::PREFIX . '_price_limit_step_' . ($i + 1));
-      $shipping_info = get_option('_' . Plugin::PREFIX . '_price_step_info_' . $i);
-      if ($i > 0) {
-        $previous_price_step = get_option('_' . Plugin::PREFIX . '_price_limit_step_' . $i);
-      }
-      if ($price_step) {
-        if ($price < $price_step && $price >= $previous_price_step) {
-          $text = $shipping_info ?: $text;
-        }
-      }
-      else {
-        $price_step = get_option('_' . Plugin::PREFIX . '_price_limit_step_' . $i);
-        if ($price_step && $shipping_info && $price > $price_step) {
-          $text = $shipping_info;
+    return self::get_product_dynamic_shipping_text($product);
+
+  }
+
+  /**
+   * Determines the first matching shipping rule.
+   *
+   * @param WC_Product $product
+   *   The product to check rules against.
+   *
+   * @return string
+   *   The shipping info text to be outputed.
+   */
+  public static function get_product_dynamic_shipping_text(\WC_Product $product): string {
+    $dynmic_shipping_rules = Admin::get_dynamic_shipping_rules();
+    $shipping_country = WC()->customer->get_shipping_country();
+    $product_shipping_class = $product->get_shipping_class();
+    $price = wc_get_price_to_display($product);
+
+    foreach ($dynmic_shipping_rules as $dynamic_rule) {
+      // Check if both product rule have no shipping class or match.
+      if ((empty($dynamic_rule['shipping_class']) && empty($product_shipping_class)) || in_array($product_shipping_class, $dynamic_rule['shipping_class'])) {
+        $sorted_shipping_rules = Plugin::get_sorted_by_price_shipping_rules($dynamic_rule['shipping_class_inner_rules']);
+        foreach ($sorted_shipping_rules as $shipping_rule) {
+          if (in_array($shipping_country, $shipping_rule['country']) && $price >= $shipping_rule['min_price']) {
+            return $shipping_rule['shipping_info'];
+          }
         }
+
       }
+
     }
-    return $text;
+    return "";
+
+  }
+
+  /**
+   * Gets shipping class inner rules sorted by price.
+   *
+   * @param array $shipping_rules
+   *   Array of Shipping class inner rules.
+   *
+   * @return array
+   *   Sorted by price array of shipping class inner rules.
+   */
+  public static function get_sorted_by_price_shipping_rules($shipping_rules) {
```

Comment:
2 things here: 
- Personally I don't think this method is worth to exist, you can do this right in the same place you call it 
- if you decide to keep it, I don't leave it in this class because I don't think it belongs in here

Reply:
I've removed the method and do the sorting the in the same place I was calling it as suggested.